### DISCUSSION TITLE
Call ctor for Collection only once

### DISCUSSION THREAD

**Discussion by mu88** [xunit/xunit#2262] [Question] [answered]
Hey guys,
I'm using the [Collection Fixture](https://xunit.net/docs/shared-context#collection-fixture) to share a `DbContext` (EF Core) across multiple test classes.
Is there any chance to call the constructor of a test class only once instead of x times (where x is number of tests) when using Collection Fixtures?
I need this to accomplish the following approach:
- Create and initialize database (within Collection Fixture) → DB template
- Create a DB copy per test class by copying the DB template. For this, I need to call the constructor of the test class exactly once.

**Comment by bradwilson** [comment, marked as answer]
Unfortunately, no. This is an important design part of xUnit.net for isolation that ensures that each test is given a fresh environment (aka, a new instance of the test class).

I assume you're calling some method on the fixture to do the "DB copy per test class" work. The simplest work around today would be to use the type of the caller as a trigger to determine when to run the logic or not. Your "get me a new copy of the database" operation can basically NOOP if the caller is the same type as the last caller.

Alternatively, you could do this as a class fixture instead of a collection fixture, and then just do the work in the class fixture's constructor, since that will only get called once per test class (but each test class that used it as its class fixture would get a fresh copy of that class, and thus a fresh copy of the database). You would still need to put such tests into a collection to prevent parallelization collisions, of course (if you have parallelization enabled).
